builtin/v13/migration/miner.go

```diff
@@ -0,0 +1,199 @@
+	var inState miner12.State
+	if err := store.Get(ctx, in.Head, &inState); err != nil {
+		return nil, xerrors.Errorf("failed to load miner state for %s: %w", in.Address, err)
+	}
+
+	ctxStore := adt.WrapStore(ctx, store)
+
+	mid, err := address.IDFromAddress(in.Address)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to get miner ID: %w", err)
+	}
+
+	inSectors, err := adt.AsArray(ctxStore, inState.Sectors, miner12.SectorsAmtBitwidth)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to load sectors array: %w", err)
+	}
+
+	hasCached, prevSectors, err := in.Cache.Read(migration.MinerPrevSectorsInKey(in.Address))
+	if err != nil {
+		return nil, xerrors.Errorf("failed to read prev sectors from cache: %w", err)
+	}
+
+	var sector miner.SectorOnChainInfo
+	if !hasCached {
+		// no cached migration, so we simply iterate all sectors and collect deal IDs
+
+		err = inSectors.ForEach(&sector, func(i int64) error {
+			if len(sector.DealIDs) == 0 || sector.Expiration < UpgradeHeight {
+				return nil
+			}
+
+			m.providerSectors.lk.Lock()
+			for _, dealID := range sector.DealIDs {
+				m.providerSectors.dealToSector[dealID] = abi.SectorID{
+					Miner:  abi.ActorID(mid),
+					Number: abi.SectorNumber(i),
+				}
+			}
+
+			sectorDeals, ok := m.providerSectors.minerToSectorToDeals[abi.ActorID(mid)]
+			if !ok {
+				sectorDeals = make(map[abi.SectorNumber][]abi.DealID)
+				m.providerSectors.minerToSectorToDeals[abi.ActorID(mid)] = sectorDeals
+			}
+			// There can't already be an entry for this sector, so just set (not append)
+			// TODO: golang: I need to copy this, since sector is a reference that's constantly updated? Right? Steb?
+			sectorDealIDs := sector.DealIDs
+			sectorDeals[abi.SectorNumber(i)] = sectorDealIDs
+
+			m.providerSectors.lk.Unlock()
+
+			return nil
+		})
+		if err != nil {
+			return nil, xerrors.Errorf("failed to iterate sectors: %w", err)
+		}
+	} else {
+		prevInSectors, err := adt.AsArray(ctxStore, prevSectors, miner12.SectorsAmtBitwidth)
+		if err != nil {
+			return nil, xerrors.Errorf("failed to load previous input sectors array: %w", err)
+		}
+
+		diffs, err := amt.Diff(ctx, store, store, prevSectors, inState.Sectors, amt.UseTreeBitWidth(miner12.SectorsAmtBitwidth))
+		if err != nil {
+			return nil, xerrors.Errorf("failed to diff old and new Sector AMTs: %w", err)
+		}
+
+		for _, change := range diffs {
+			sectorNo := abi.SectorNumber(change.Key)
+
+			switch change.Type {
+			case amt.Add:
+				fallthrough
+			case amt.Modify:
+				found, err := inSectors.Get(change.Key, &sector)
+				if err != nil {
+					return nil, xerrors.Errorf("failed to get sector %d in inSectors: %w", sectorNo, err)
+				}
+
+				if !found {
+					return nil, xerrors.Errorf("didn't find sector %d in inSectors", sectorNo)
+				}
+
+				if len(sector.DealIDs) == 0 {
```

Comment:
Modify needs to be treated as remove + add, so we need to run the remove logic from below even in this case.